### PR TITLE
Add support for Helm chart linting and releasing.

### PR DESCRIPTION
**Pull request by mprimeaux**
### Description
* Added support for the Helm chart testing action (linting and testing).
* Added support for the Helm chart releaser action.
* Fixed minor lint issues in Helm chart values files.
* Added GitHub Actions Workflow badges in README.
  
### Issues Resolved
[#14](https://github.com/opensearch-project/helm-charts/issues/14)
 
### Check List
- :white_check_mark: Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by mprimeaux** [comment]
@peterzhuamazon @TheAlgo @DandyDeveloper @smlx  I set the GH pages branch as `repository` and not the default of `gh-pages`.  Please review and let me know if any changes are required. All comments and feedback are welcome.

File: .github/workflows/lint-test.yaml
Comment (DandyDeveloper) [inline .github/workflows/lint-test.yaml:40]: I assume something is wrong with this because of helm 3? I found a similar problem on other repos.

File: .github/workflows/lint-test.yaml
Comment (DandyDeveloper) [inline .github/workflows/lint-test.yaml:51]: This missing kind of defeats a core part of CI testing on the chart. 

What's the reason for this being removed?

File: .github/workflows/release.yaml
Comment (DandyDeveloper) [inline .github/workflows/release.yaml:35]: @TheAlgo Someone with settings access will need to create this.

File: .gitignore
Comment (DandyDeveloper) [inline .gitignore:3]: Redundant? Maybe IDE specific?

File: charts/opensearch-dashboards/values.yaml
Comment (DandyDeveloper) [inline charts/opensearch-dashboards/values.yaml:62]: Two ## better indicate a command. Is this a linting thing?

**Review by DandyDeveloper — COMMENTED** [review]
Looks good for the most part, but I have some comments.

File: .github/workflows/lint-test.yaml
Comment (mprimeaux) [reply .github/workflows/lint-test.yaml:40]: The issue is not with helm 3 but with the fact that the GitHub runner only has 2 CPUs and 8 GB memory, which causes a failure with the default value of `replicas: 3`.  

The attached [log](https://github.com/opensearch-project/helm-charts/files/7171243/ct-install-log.txt) shows that `ct install` succeeds in testing both charts when using Minikube on my local machine with enough resources.

I can confirm that using `replicas: 1` in the values file for the `opensearch` chart within the GitHub runner succeeds.

The Kind GitHub action is fairly limited in what we can do with it and so I switched to the Minikube GitHub action. Setting `--cpus 2 --memory 6144mb`.  What fails is a replica count greater than 1 **but only within the GitHub runner**.

I do agree that we want to test and not only lint. I propose we (1) keep the `replicas: 3` in the values file and (2) override it to `1` in the GitHub actions workflow, which does provide some level of testing. 

What are your thoughts?

File: .github/workflows/lint-test.yaml
Comment (mprimeaux) [reply .github/workflows/lint-test.yaml:51]: I included a reply to this comment above.

File: .github/workflows/release.yaml
Comment (mprimeaux) [reply .github/workflows/release.yaml:35]: There is no need to create this as it's [provided by GitHub](https://docs.github.com/en/actions/reference/authentication-in-a-workflow).

File: .gitignore
Comment (mprimeaux) [reply .gitignore:3]: This is for macOS and unrelated to an IDE. macOS creates this file when you visit a folder in Finder. It is needed to avoid adding it inadvertently.

File: charts/opensearch-dashboards/values.yaml
Comment (mprimeaux) [reply charts/opensearch-dashboards/values.yaml:62]: Yes, it's a linting thing.

File: .gitignore
Comment (peterzhuamazon) [reply .gitignore:3]: I want to raise the bar and see if we can add all these:
```
Thumbs.db
.DS_Store

.idea/
*.iml

__pycache__
.pytest_cache

.coverage
coverage.xml

.gradle/
.settings/
out.txt
```

Comment (peterzhuamazon) [reply .gitignore:3]: This is the standard one we use in opensearch-build repo.

Comment (mprimeaux) [reply .gitignore:3]: Agreed. I'll add the above.

Comment (mprimeaux) [reply .gitignore:3]: If we really want to 'go wild' then there is [this](https://github.com/github/gitignore).

File: .github/workflows/lint-test.yaml
Comment (mprimeaux) [reply .github/workflows/lint-test.yaml:40]: After a bit of testing various options, we have a few options:

**Option 1:** Leave `replicas: 3` and use an external GitHub runner. 

I don't really like this option.

**Option 2:** Set `replicas: 1` and include a note indicating the value should not be set to a value of less than 3 for production operational environments.

This is also a less than ideal option because, by default, this results in a single copy of the index. So folks would almost always need to override this value. But this will allow tests to pass.

The reason why this option is 'less than ideal' is because OpenSearch sets the index replica count to 1 and there's a rule that disallows the index replica to be deployed to the same pod (a.k.a. OpenSearch node).

**Option 3:** Leave `replicas: 3` and 'fix-up' the `charts/opensearch/values.yaml` file **before** running chart testing.

I think this is our best option as it tests the chart and keeps minimum operating values 'as is'. In the end, we aren't testing OpenSearch but are testing the chart.

Comment (mprimeaux) [reply .github/workflows/lint-test.yaml:40]: For the curious, [here](https://github.com/mprimeaux/helm-charts/runs/3613235595?check_suite_focus=true) is a successful run with linting and testing.  I simply used `sed` in a custom action to modify the values file.

File: charts/opensearch-dashboards/values.yaml
Comment (DandyDeveloper) [reply charts/opensearch-dashboards/values.yaml:62]: Space wasn't the concern here, you removed ## -> #. I'm wondering if that part is intentional?

File: .gitignore
Comment (DandyDeveloper) [reply .gitignore:3]: I didn't know MacOS added those default, sorry!

File: .github/workflows/release.yaml
Comment (DandyDeveloper) [reply .github/workflows/release.yaml:35]: Weird, I swear back in the day I had to make this. Thank you for the link!

File: charts/opensearch-dashboards/values.yaml
Comment (mprimeaux) [reply charts/opensearch-dashboards/values.yaml:62]: The `##` needed to be `# ` to avoid the yaml lint error.

File: .github/workflows/lint-test.yaml
Comment (peterzhuamazon) [reply .github/workflows/lint-test.yaml:40]: We have tried option 1 back in ODFE and it is painful to maintain a provision script, bootstrap to GitHub code, and the cleanup code. Sometimes the cleanup is not working and we have to manually delete the runner in GitHub.

I like Option 3 and agree with your comments. We are testing if charts work.

Comment (DandyDeveloper) [reply .github/workflows/lint-test.yaml:40]: By fix-up, we mean override with something like `--set replicas=2`? 

I don't see it in the step: https://github.com/mprimeaux/helm-charts/runs/3613235595?check_suite_focus=true#step:10:30 

Maybe I'm misunderstanding. Sorry.

Comment (smlx) [reply .github/workflows/lint-test.yaml:40]: you can override the defaults in CI by creating a `ci/ci-values.yaml` file inside the chart.

> Charts may have multiple custom values files matching the glob pattern '*-values.yaml' in a directory named 'ci' in the root of the chart's directory. The chart is installed and tested for each of these files. If no custom values file is present, the chart is installed and tested with defaults.

From https://github.com/helm/chart-testing/blob/main/doc/ct_install.md

Comment (smlx) [reply .github/workflows/lint-test.yaml:40]: I'd recommend putting values in here to scale CI to be as small and fast as possible: no persistence, single replica etc.

Comment (smlx) [reply .github/workflows/lint-test.yaml:40]: > The Kind GitHub action is fairly limited in what we can do with it and so I switched to the Minikube GitHub action.

I'm quite familiar with `kind` and the github action. What issues did you run into here?

The nice thing about `kind` is the speed compared to `minikube`.

Comment (mprimeaux) [reply .github/workflows/lint-test.yaml:40]: @smlx Thanks for pointing out the `ci` folder. I didn't see that option when reviewing the `ct install` command. I do think this is the best option.

Regarding the use of Minikube over Kind, I should have clarified that I switched to Minikube in an attempt to use all resources (cpu and memory) available on the GitHub runner by adding the following step:

```
# See https://github.com/marketplace/actions/setup-minikube-kubernetes-cluster
- name: Create minikube cluster
  uses: manusa/actions-setup-minikube@v2.4.2
  if: steps.list-changed.outputs.changed == 'true'
  with:
    minikube version: 'v1.23.0'
    kubernetes version: 'v1.21.4'
    driver: docker
    github token: ${{ secrets.GITHUB_TOKEN }}
    start args: '--addons metrics-server --cpus 2 --memory 6144mb'
```

This isn't needed anymore since `replicas: 1` will be set in the CI-specific values file.

Comment (mprimeaux) [reply .github/workflows/lint-test.yaml:40]: Hi @DandyDeveloper 

> By fix-up, we mean override with something like `--set replicas=2`?
> 
> I don't see it in the step: https://github.com/mprimeaux/helm-charts/runs/3613235595?check_suite_focus=true#step:10:30
> 
> Maybe I'm misunderstanding. Sorry.

I had created the following step before running `ct install`:

```
- name: Fixup values file
  run: |
    sed -i 's/replicas: 3/replicas: 1/g' charts/opensearch/values.yaml
```

Based on what @smlx pointed out, this step of course isn't needed in lieu of the CI-specific values file.

Comment (mprimeaux) [reply .github/workflows/lint-test.yaml:40]: Hi @smlx 

> The nice thing about `kind` is the speed compared to `minikube`.

FWIW, the duration difference between Minikube and Kind for their respective steps was marginal:

* Kind: [1 min 49 seconds](https://github.com/mprimeaux/helm-charts/runs/3621760978?check_suite_focus=true)
* Minikube: [1 min 14 seconds](https://github.com/mprimeaux/helm-charts/runs/3613235595?check_suite_focus=true)

Nevertheless, as suggested, I reverted to using Kind.

**Comment by mprimeaux** [comment]
@smlx @DandyDeveloper @TheAlgo @peterzhuamazon I believe I've addressed everyone's feedback and sincerely appreciate the time you all took to review. 

Please let me know if there are any other concerns or feedback.

File: .github/workflows/lint-test.yaml
Comment (TheAlgo) [reply .github/workflows/lint-test.yaml]: I also kinda agree on using `kind` only. Thanks @mprimeaux for this.

**Review by TheAlgo — APPROVED** [review]
Changes looks good to me.

File: .github/workflows/lint-test.yaml
Comment (peterzhuamazon) [reply .github/workflows/lint-test.yaml]: Kind +1

**Review by DandyDeveloper — APPROVED** [review]
Thank you for clarifying everything. Great start.

**Comment by TheAlgo** [comment]
@smlx @peterzhuamazon Can you review the latest changes as per the feedback. We can push this PR after that

File: .github/workflows/lint-test.yaml
Comment (smlx) [inline .github/workflows/lint-test.yaml:25]: This is pretty old, could we bump it to the latest version?

File: .github/workflows/lint-test.yaml
Comment (smlx) [inline .github/workflows/lint-test.yaml:11]: I'm not sure what the workflow on this repository is, but if `main` is protected from pushes and PRs are required then this can be just
```
on: pull_request
```

File: .github/workflows/lint-test.yaml
Comment (smlx) [inline .github/workflows/lint-test.yaml:15]: minor nit, but `ubuntu-latest` means this doesn't need to be bumped every few years :slightly_smiling_face:

File: .gitignore
Comment (smlx) [inline .gitignore:1]: This file doesn't belong in the repository, instead it should go in your system git configuration.

https://git-scm.com/docs/gitignore

File: cr.yaml
Comment (smlx) [inline cr.yaml:1]: I think this file is unnecessary since these values are automatically detected by the action: https://github.com/helm/chart-releaser-action/blob/main/action.yml#L21-L22

And the `pages-branch` is `gh-pages` by default, which matches the github pages default branch which is used as the helm repository. Any reason why it is not `gh-pages` here?

File: ct.yaml
Comment (smlx) [inline ct.yaml:11]: IMO all the checks should be enabled. Any other opinions?

**Review by smlx — CHANGES_REQUESTED** [review]
A few comments and questions but a great start overall.

File: .github/workflows/lint-test.yaml
Comment (mprimeaux) [reply .github/workflows/lint-test.yaml:25]: Sure.

File: .github/workflows/lint-test.yaml
Comment (mprimeaux) [reply .github/workflows/lint-test.yaml:15]: Sure. I'll modify it.

I had pinned it due to the lag when `latest` still resulted in Ubuntu 18.04 LTS even though Ubuntu 20.04 LTS was released. It was a bit frustrating since `latest` didn't actually mean latest.

File: .gitignore
Comment (mprimeaux) [reply .gitignore:1]: Thanks. I wasn't aware of that subtle difference.

File: cr.yaml
Comment (mprimeaux) [reply cr.yaml:1]: It's a personal preference to convey more semantic meaning rather than the general name of `gh-pages`. Happy to change since there is enough context given this repository is solely dedicated to Helm charts.

File: ct.yaml
Comment (mprimeaux) [reply ct.yaml:11]: I agree.

File: .github/workflows/lint-test.yaml
Comment (mprimeaux) [reply .github/workflows/lint-test.yaml:11]: I wasn't sure of the workflow, either. The intention is to lint and test everywhere except the `main` and `gh-pages` (formerly `repository`) branches. 

I'll assume `on: pull_request` is correct as per your suggestion.

**Comment by mprimeaux** [comment]
@smlx I believe your comments have been addressed. Please let me know.

**Comment by peterzhuamazon** [comment]
I have no objections, will let @smlx to confirm the changes from @mprimeaux before merge.
Thanks everyone.

File: .github/workflows/lint-test.yaml
Comment (peterzhuamazon) [reply .github/workflows/lint-test.yaml]: I personally like to use specific number as latest can change and some code does not work the same way between OS versions sometimes. This one tho, I am not too worried to use latest as it is just a lint test.

File: README.md
Comment (TheAlgo) [inline README.md:15]: Sorry to bring it now, but can we not add a title for installation here,

Comment (mprimeaux) [reply README.md:17]: Done.

**Comment by mprimeaux** [comment]
@smlx Please let me know if any other concerns need to be addressed. We are hoping to leverage this PR in our release this week. Your support is appreciated.

**Comment by mprimeaux** [comment]
I am just checking in to see if there are any other outstanding changes being requested.  Thanks much.

**Comment by peterzhuamazon** [comment]
Thanks @mprimeaux I see you have addressed all findings, I am also waiting for @smlx to final confirm.
We will check back again tomorrow, and will merge this if no more comments.


Thanks.

**Comment by peterzhuamazon** [comment]
Not seeing any more comments will approve now.
Thanks.

**Pull request merged**
